src/Features/PlayerTrace.hpp

```diff
@@ -1,27 +1,56 @@
 #pragma once
 
+#include <map>
+
 #include "Feature.hpp"
 #include "Utils.hpp"
 #include "Features/Hud/Hud.hpp"
 
-class PlayerTrace : public Feature {
-private:
-    std::vector<Vector> trace;
-    // Records at which indexes into the trace the grounded state changes
-    std::vector<size_t> groundframe_swaps;
-    // Records whether player started grounded or not
-    bool started_grounded;
+#define TRACE_SCALE 10
+// Stores a position delta as a fixed point number
+struct TraceDelta {
+    int16_t dx : 10;
+    int16_t dy : 10;
+    int16_t dz : 10;
+
+    inline TraceDelta(const Vector vect): dx(vect.x), dy(vect.y), dz(vect.z) {}
+    inline Vector asVector() const {
+        return Vector(dx, dy, dz);
+    }
+};
 
-    void DrawSpeedDelta(HudContext *ctx, int *hud_id, size_t begin_tick, size_t end_tick) const;
+// Stores accurate enough info about the player for the trace
+// Tick is an index into the position deltas
+struct TraceUpdate {
+    size_t tick;
+    float speed;
+    int32_t x : 21;
+    int32_t y : 21;
+    int32_t z : 21;
+    bool grounded : 1;
+};
+
+struct Trace {
+    // Keeps track of player pos
+    // The interval between them is not constant
+    std::vector<TraceDelta> deltas;
+    // Keeps track of groudframes, speed at important moments in the hops
+    std::vector<TraceUpdate> updates;
+    // Remember the last pos, in order to compute deltas
+    // Note: it is actually position times TRACE_SCALE
+    Vector last_pos;
+};
 
+class PlayerTrace : public Feature {
+private:
+    // In order to arbitrarily number traces
+    std::map<size_t, Trace> traces;
 public:
     PlayerTrace();
     // Add a point to the player trace
-    void AddPoint(const Vector Point);
+    void AddPoint(const size_t trace_idx, void *player);
```

Comment:
a plain (non-pointer, non-reference) argument doesn't really need to be const-qualified, i honestly didn't even know you COULD do that